pkg/job/job.go

```diff
@@ -0,0 +1,252 @@
+
+	"github.com/MinhNghiaD/jobworker/api/worker/proto"
+	"github.com/MinhNghiaD/jobworker/pkg/log"
+)
+
+// Job is managed by the worker for the execution of linux process
+type Job interface {
+	// Start the job in the background
+	Start() error
+	// Stop the job
+	Stop(force bool) error
+	// Query the current status of the job
+	Status() *proto.JobStatus
+}
+
+// TODO Handle QUEUING state when job queue and scheduler is implemented
+// For now, the job will be start immediately after created
+// Process waiting to be executed
+
+// Impl is the implementation of the Job interface
+type Impl struct {
+	id         uuid.UUID
+	cmd        *exec.Cmd
+	logger     *log.Logger
+	owner      string
+	state      proto.ProcessState
+	stateMutex sync.RWMutex
+	exitChan   chan bool
+}
+
+// newJob creates a new job, associated with a unique ID and a Command, and the owner who requests its
+func newJob(ID uuid.UUID, cmd *exec.Cmd, logger *log.Logger, owner string) (Job, error) {
+	return &Impl{
+		id:       ID,
+		cmd:      cmd,
+		logger:   logger,
+		owner:    owner,
+		state:    proto.ProcessState_RUNNING,
+		exitChan: make(chan bool, 1),
+	}, nil
+}
+
+// Start runs job in the background
+func (j *Impl) Start() error {
+	// IO Mapping
+	stdoutLog := j.logger.Entry.WithField("source", "stdout").Writer()
+	stderrLog := j.logger.Entry.WithField("source", "stderr").Writer()
+
+	j.cmd.Stdout = stdoutLog
+	j.cmd.Stderr = stderrLog
+	j.cmd.Stdin = nil
+
+	logrus.Infof("Starting job %s", j.id)
+
+	err := j.cmd.Start()
+	if err != nil {
+		j.changeState(proto.ProcessState_STOPPED)
+
+		badRequest := &errdetails.BadRequest{
+			FieldViolations: []*errdetails.BadRequest_FieldViolation{
+				{
+					Field:       "command",
+					Description: err.Error(),
+				},
+			},
+		}
+
+		return ReportError(codes.PermissionDenied, "Fail to start job", badRequest)
+	}
+
+	go func() {
+		if err := j.cmd.Wait(); err != nil {
+			j.logger.Entry.Infof("Job %s terminated, reason %s", j.id, err)
+		}
+
+		// Close logger
+		if err := stdoutLog.Close(); err != nil {
+			logrus.Errorln(err)
+		}
+
+		if err := stderrLog.Close(); err != nil {
+			logrus.Errorln(err)
+		}
+
+		j.logger.Entry.Infof("Exiting job %s", j.id)
+
+		if err := j.logger.Close(); err != nil {
+			logrus.Errorln(err)
+		}
+
+		j.changeState(proto.ProcessState_EXITED)
+
+		// Inform stop caller to finish.
+		// The channel buffer size is 1, so if this job exit naturally, it won't be blocked
+		j.exitChan <- true
+		close(j.exitChan)
+	}()
+
+	return nil
+}
+
+// Stop terminates a job. If the force flag is set to true, it will using SIGKILL to terminate the process, otherwise it will be SIGTERM
+func (j *Impl) Stop(force bool) error {
+	if j.getState() != proto.ProcessState_RUNNING || j.cmd.Process == nil {
+		badRequest := &errdetails.BadRequest{
+			FieldViolations: []*errdetails.BadRequest_FieldViolation{
+				{
+					Field:       "job",
+					Description: "Job not running",
+				},
+			},
+		}
```

Comment:
Can you move all of these to a helper function? All of these seem to be the returning the same struct with slight difference in description, maybe create a function called `func createError(description string) *errdetails.BadRequest`. As it is, it's quite hard to read.

Reply:
Yes I will try to reorganize them into the same pattern

Reply:
It is corrected